### PR TITLE
Add a tsdbutil command to analyse churn etc.

### PR DESCRIPTION
**Pull request by brian-brazil**
This reports the cardinality of each label,
the total number of label pairs,
and how much series worth of time is "uncovered"
by series data. Which is basically how much churn there is.

Still needs UI work, but the core is there.

### CODE REVIEW

File: cmd/tsdb/main.go [hunk 1]
```diff
@@ -383,3 +396,89 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
...
+	fmt.Printf("Postings (unique label pairs): %d\n", len(labelpairsUncovered))
+	fmt.Printf("Postings entries: %d\n", entries)
+
+	postingInfos = postingInfos[:0]
+	for k, m := range labelpairsCount {
+		postingInfos = append(postingInfos, postingInfo{k, m})
+	}
+
+	fmt.Printf("\nMost Common Label Pairs:\n")
+	printInfo()
```
Comment (krasi-georgiev) [inline cmd/tsdb/main.go]: probably best to have this as `printInfo(postingInfos)` to be a bit more explicit.

File: cmd/tsdb/main.go [hunk 2]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
 	}
 	return strconv.FormatInt(timestamp, 10)
 }
+
+func analyzeBlock(b *tsdb.Block, limit int) {
+	fmt.Printf("Block Path: %s\n", b.Dir())
+	meta := b.Meta()
+	// Presume 1ms resoluation that Prometheus uses.
```
Comment (juliusv) [inline cmd/tsdb/main.go]: resoluation -> resolution :P

File: cmd/tsdb/main.go [hunk 3]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
 	}
 	return strconv.FormatInt(timestamp, 10)
 }
+
+func analyzeBlock(b *tsdb.Block, limit int) {
+	fmt.Printf("Block Path: %s\n", b.Dir())
```
Comment (juliusv) [inline cmd/tsdb/main.go]: "Path" -> "path"?

File: cmd/tsdb/main.go [hunk 4]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
...
+	meta := b.Meta()
+	// Presume 1ms resoluation that Prometheus uses.
+	fmt.Printf("Duration: %s\n", (time.Duration(meta.MaxTime-meta.MinTime) * 1000000).String())
```
Comment (juliusv) [inline cmd/tsdb/main.go]: IMO `1e6` is more readable / hard to get wrong than `1000000`.

File: cmd/tsdb/main.go [hunk 5]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
...
+	if err != nil {
+		exitWithError(err)
+	}
+	defer ir.Close()
+
+	ln, err := ir.LabelNames()
+	if err != nil {
+		exitWithError(err)
+	}
+	fmt.Printf("LabelNames: %d\n", len(ln))
```
Comment (juliusv) [inline cmd/tsdb/main.go]: In user output, "Label names:" would make more sense?

File: cmd/tsdb/main.go [hunk 6]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
...
+
+	fmt.Printf("\nLabel Pairs most involved in Churning:\n")
+	printInfo(postingInfos)
+
+	postingInfos = postingInfos[:0]
+	for k, m := range labelsUncovered {
+		postingInfos = append(postingInfos, postingInfo{k, uint64(float64(m) / float64(meta.MaxTime-meta.MinTime))})
+	}
+
+	fmt.Printf("\nLabel Names most involved in Churning:\n")
```
Comment (juliusv) [inline cmd/tsdb/main.go]: Names -> names
Churning -> churning

File: cmd/tsdb/main.go [hunk 7]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
...
+	}
+	fmt.Printf("Postings (unique label pairs): %d\n", len(labelpairsUncovered))
+	fmt.Printf("Postings entries (total label pairs): %d\n", entries)
+
+	postingInfos = postingInfos[:0]
+	for k, m := range labelpairsUncovered {
+		postingInfos = append(postingInfos, postingInfo{k, uint64(float64(m) / float64(meta.MaxTime-meta.MinTime))})
+	}
+
+	fmt.Printf("\nLabel Pairs most involved in Churning:\n")
```
Comment (juliusv) [inline cmd/tsdb/main.go]: Pairs -> pairs
Churning -> churning

File: cmd/tsdb/main.go [hunk 8]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
...
+
+	fmt.Printf("\nLabel Names most involved in Churning:\n")
+	printInfo(postingInfos)
+
+	postingInfos = postingInfos[:0]
+	for k, m := range labelpairsCount {
+		postingInfos = append(postingInfos, postingInfo{k, m})
+	}
+
+	fmt.Printf("\nMost Common Label Pairs:\n")
```
Comment (juliusv) [inline cmd/tsdb/main.go]: similar (and below)...

**Comment by brian-brazil** [comment]
All done.

File: cmd/tsdb/main.go [hunk 9]
```diff
@@ -64,6 +69,27 @@ func main() {
...
+		db, err := tsdb.Open(*analyzePath, nil, nil, nil)
+		if err != nil {
+			exitWithError(err)
+		}
+		blocks := db.Blocks()
+		var block *tsdb.Block
+		if *analyzeBlockId != "" {
+			for _, b := range blocks {
+				if b.Meta().ULID.String() == *analyzeBlockId {
+					block = b
```
Comment (codesome) [inline cmd/tsdb/main.go]: We can `break` after this.

File: cmd/tsdb/main.go [hunk 10]
```diff
@@ -64,6 +69,27 @@ func main() {
...
+				}
+			}
+
```
Comment (codesome) [inline cmd/tsdb/main.go]:
```suggestion
```

File: cmd/tsdb/main.go [hunk 11]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
 	}
 	return strconv.FormatInt(timestamp, 10)
 }
+
+func analyzeBlock(b *tsdb.Block, limit int) {
```
Comment (codesome) [inline cmd/tsdb/main.go]: Can we have single line comment at the beginning of each subsection of the code about what it is trying to calculate (in addition to `Printf` at the ends)? (makes the code easier to follow)

File: cmd/tsdb/main.go [hunk 12]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
...
+	printInfo(postingInfos)
+
+	postingInfos = postingInfos[:0]
+	for k, m := range labelpairsCount {
+		postingInfos = append(postingInfos, postingInfo{k, m})
+	}
+
+	fmt.Printf("\nMost common label pairs:\n")
+	printInfo(postingInfos)
+
```
Comment (krasi-georgiev) [inline cmd/tsdb/main.go]: postingInfos should be  reset here:

`postingInfos = postingInfos[:0]`

File: cmd/tsdb/main.go [hunk 13]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
...
+	// Presume 1ms resolution that Prometheus uses.
+	fmt.Printf("Duration: %s\n", (time.Duration(meta.MaxTime-meta.MinTime) * 1e6).String())
+	fmt.Printf("Series: %d\n", meta.Stats.NumSeries)
+	ir, err := b.Index()
+	if err != nil {
+		exitWithError(err)
+	}
+	defer ir.Close()
+
+	ln, err := ir.LabelNames()
```
Comment (krasi-georgiev) [inline cmd/tsdb/main.go]: since this is used quite a bit further down lets have it as `allLabelNames`

File: cmd/tsdb/main.go [hunk 14]
```diff
@@ -383,3 +409,129 @@ func getFormatedTime(timestamp int64, humanReadable *bool) string {
...
+
+	postingInfos = postingInfos[:0]
+	for k, m := range labelpairsCount {
+		postingInfos = append(postingInfos, postingInfo{k, m})
+	}
+
+	fmt.Printf("\nMost common label pairs:\n")
+	printInfo(postingInfos)
+
+	for _, n := range ln {
```
Comment (krasi-georgiev) [inline cmd/tsdb/main.go]: postingInfos should be  reset before this line:

`postingInfos = postingInfos[:0]`

File: cmd/tsdb/main.go (hunk 11 above)
Comment (brian-brazil) [reply cmd/tsdb/main.go]: I'm not sure that'd help much tbh

**Comment by brian-brazil** [comment]
I think that's everything addresses.

**Comment by krasi-georgiev** [comment]
@brian-brazil The code doesn't account for deleted data in the tombstone file, isn't this important?

**Comment by krasi-georgiev** [comment]
maybe also add an entry in the changelog file so that this is also included in the Prometheus changelog and the other maintainers/users know about it and start using it.

**Comment by brian-brazil** [comment]
> The code doesn't account for deleted data in the tombstone file, isn't this important?

Not really, as long as the data is in the index it's going to slow things down and use up RAM.

**Pull request merged**